snapshots/manager.go

```diff
@@ -0,0 +1,223 @@
+	m.operation = opNone
+	if m.chRestore != nil {
+		close(m.chRestore)
+		m.chRestore = nil
+	}
+	m.chRestoreDone = nil
+	m.restorePending = nil
+}
+
+// List lists snapshots, mirroring ABCI ListSnapshots. It can be concurrent with other operations.
+func (m *Manager) List() ([]*Snapshot, error) {
+	return m.store.List()
+}
+
+// LoadChunk loads a chunk into a byte slice, mirroring ABCI LoadChunk. It can be called
+// concurrently with other operations. If the chunk does not exist, nil is returned.
+func (m *Manager) LoadChunk(height uint64, format uint32, chunk uint32) ([]byte, error) {
+	reader, err := m.store.LoadChunk(height, format, chunk)
+	if err != nil {
+		return nil, err
+	}
+	if reader == nil {
+		return nil, nil
+	}
+	defer reader.Close()
+
+	return ioutil.ReadAll(reader)
+}
+
+// Take takes a snapshot and returns its metadata.
+func (m *Manager) Take(height uint64) (*Snapshot, error) {
+	if m == nil {
+		return nil, errors.New("no snapshot store configured")
+	}
+	err := m.begin(opSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	defer m.end()
+
+	latest, err := m.store.GetLatest()
+	if err != nil {
+		return nil, fmt.Errorf("failed to examine latest snapshot: %w", err)
+	}
+	if latest != nil && latest.Height >= height {
+		return nil, fmt.Errorf("a more recent snapshot already exists at height %v", latest.Height)
+	}
+
+	chunks, err := m.target.Snapshot(height, CurrentFormat)
+	if err != nil {
+		return nil, err
+	}
+	return m.store.Save(height, CurrentFormat, chunks)
+}
+
+// Prune prunes snapshots, if no other operations are in progress.
+func (m *Manager) Prune(retain uint32) (uint64, error) {
+	err := m.begin(opPrune)
+	if err != nil {
+		return 0, err
+	}
+	defer m.end()
+	return m.store.Prune(retain)
+}
+
+// Restore begins an async snapshot restoration, mirroring ABCI OfferSnapshot. Chunks must be fed
+// via RestoreChunk() until the restore is complete or a chunk fails.
+func (m *Manager) Restore(snapshot Snapshot) error {
+	if snapshot.Chunks == 0 {
+		return fmt.Errorf("%w: no chunks", ErrInvalidMetadata)
+	}
+	if uint32(len(snapshot.Metadata.ChunkHashes)) != snapshot.Chunks {
+		return fmt.Errorf("%w: snapshot has %v chunk hashes, but %v chunks",
+			ErrInvalidMetadata,
+			uint32(len(snapshot.Metadata.ChunkHashes)),
+			snapshot.Chunks)
+	}
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	err := m.beginLocked(opRestore)
+	if err != nil {
+		return err
+	}
+
+	// Start an asynchronous snapshot restoration, passing chunks and completion status via channels.
+	chChunks := make(chan io.ReadCloser, chunkBufferSize)
+	chDone := make(chan error, 1)
+	go func() {
+		chDone <- m.target.Restore(snapshot.Height, snapshot.Format, chChunks)
+		close(chDone)
+	}()
+
+	// Check for any initial errors from the restore, before any chunks are fed.
+	select {
+	case err := <-chDone:
+		if err == nil {
+			err = errors.New("restore ended unexpectedly")
+		}
+		m.endLocked()
+		return err
+	case <-time.After(20 * time.Millisecond):
```

Comment:
why `time.After` is added?

Reply:
We need to know if the goroutine above (`m.target.Restore()`) had an error, so we have to wait for it to be scheduled. This is a real code smell, so if you have a better suggestion I'm all ears. One option would be for `Restore()` to return a `Restorer` struct that we could feed chunks to, and error if the snapshot is invalid, but that seemed a bit much.

Reply:
Other idea is to create another channel or a semaphore and pass it to the `Restore` function. Once it will be ready for listening, we will send a signal (channel) or open the semaphore. I think channel is easier here because you have a switch.

Reply:
Yeah, I added a ready channel:

```go
Restore(height uint64, format uint32, chunks <-chan io.ReadCloser, ready chan<- struct{}) error
```

It's not super-clean, but it works and is fairly lightweight. Would possibly be better to return a restorer, in the same way as IAVL's `MutableTree.Import()` returns an importer, but it makes the interface harder to implement and I'd like to keep it lightweight.